lib/sessions.js

```diff
@@ -243,6 +244,130 @@ class ClientSession extends EventEmitter {
   toBSON() {
     throw new Error('ClientSession cannot be serialized to BSON.');
   }
+
+  /**
+   * A user provided function to be run within a transaction
+   *
+   * @callback WithTransactionCallback
+   * @param {ClientSession} session The parent session of the transaction running the operation. This should be passed into each operation within the lambda.
+   * @returns {Promise} The resulting Promise of operations run within this transaction
+   */
+
+  /**
+   * Runs a provided lambda within a transaction, retrying either the commit operation
+   * or entire transaction as needed (and when the error permits) to better ensure that
+   * the transaction can complete successfully.
+   *
+   * IMPORTANT: This method requires the user to return a Promise, all lambdas that do not
+   * return a Promise will result in undefined behavior.
+   *
+   * @param {WithTransactionCallback} fn
+   * @param {TransactionOptions} [options] Optional settings for the transaction
+   */
+  withTransaction(fn, options) {
+    const startTime = Date.now();
+    return attemptTransaction(this, startTime, fn, options);
+  }
+}
+
+const MAX_WITH_TRANSACTION_TIMEOUT = 120000;
+const WRITE_CONCERN_FAILED_CODE = 64;
+const UNSATISFIABLE_WRITE_CONCERN_CODE = 100;
+const UNKNOWN_REPL_WRITE_CONCERN_CODE = 79;
+const NON_DETERMINISTIC_WRITE_CONCERN_ERRORS = [
+  'CannotSatisfyWriteConcern',
+  'UnknownReplWriteConcern',
+  'UnsatisfiableWriteConcern'
+];
+
+function hasNotTimedOut(startTime, max) {
+  return Date.now() - startTime < max;
+}
+
+function isWriteConcernTimeoutError(err) {
+  return err.code === WRITE_CONCERN_FAILED_CODE && !!(err.errInfo && err.errInfo.wtimeout === true);
+}
+
+function isUnknownTransactionCommitResult(err) {
+  return (
+    NON_DETERMINISTIC_WRITE_CONCERN_ERRORS.indexOf(err.codeName) === -1 &&
```

Comment:
To follow up on https://github.com/mongodb-js/mongodb-core/pull/399#discussion_r253275147:

It is more semantically correct to use a `Set` over an `Array` when checking if something exists. It is clearer to have a `Set` and say `!set.has(item)` then it is to do an `indexOf` check (as well as being less verbose)

As for hoisting it out, it is more a verbosity issue than a performance issue. It is clearer to see:

```
NON_DETERMINISTIC_WRITE_CONCERN_ERRORS.has(err.codeName);
```

instead of an inline list. And should we add more error conditions, it will be more maintainable.

Overall, I'd like to stop using Arrays when we could be using Sets.

Reply:
👍 switched to using `Set`